### PR TITLE
fix: add legacy messages back to codec

### PR DESCRIPTION
**Pull request by lumtis**
# Description

- Add old field in `VoteGasPrice` back in the message
- Define pre-v19 messages back to add in codec

Closes: https://github.com/zeta-chain/node/issues/2835

## Summary by CodeRabbit

- **New Features**
	- Introduced support for stateful precompiled contracts and enhancements for Bitcoin chain support, including multiple chains and testnet capabilities.
	- Added staking precompiled contract with event emission and support for restricted addresses in Solana.
	- Expanded cross-chain transaction message definitions with new fields for improved data capture.

- **Bug Fixes**
	- Improved outbound tracker functionality and fixed issues with Bitcoin transaction processing.

- **Tests**
	- Expanded end-to-end testing for new features, including staking precompiled contracts and Bitcoin deposit functionalities.

- **Documentation**
	- Updated changelog and message definitions to reflect recent changes and enhancements.

### CODE REVIEW

File: x/crosschain/types/legacy_msgs.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	return []sdk.AccAddress{creator}
+}
+
+func (msg *MsgVoteOnObservedInboundTx) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgVoteOnObservedInboundTx) ValidateBasic() error {
+	return nil
```
Comment (kingpinXD) [inline x/crosschain/types/legacy_msgs.go:35]: Should we return `err not supported` here ?
I am guess we don't need to execute the messages to be able to parse them on the indexer (might be wrong tough)

Comment (lumtis) [reply x/crosschain/types/legacy_msgs.go:35]: It's just implemented to implement the Msg interface. The message can't be processed if not added explicitly in the MsgService so I think it's fine

**Comment by CryptoFewka** [comment]
@kingpinXD & @lumtis, 

I've set up a basic Athens full node from a v19 snapshot, and ran it to the v20 upgrade. At the v20 upgrade height, I used a binary built from https://github.com/zeta-chain/node/tree/v20-restore-old-messages. That branch was based on `release/v20` and I cherry picked in the commits from https://github.com/zeta-chain/node/pull/2909.

The v20 flavor of the binary I produced runs, though the node only has blocks from 6711001 onward, which is after the v19 upgrade and presumably is missing knowledge of the blocks which contains the messages we need to test.

If needed, we can run this v20 binary with a copy of an archive node's data. 

Were there any messages present in v19 but not v20 that could be used to test?

**Comment by lumtis** [comment]
> @kingpinXD & @lumtis,
> 
> I've set up a basic Athens full node from a v19 snapshot, and ran it to the v20 upgrade. At the v20 upgrade height, I used a binary built from https://github.com/zeta-chain/node/tree/v20-restore-old-messages. That branch was based on `release/v20` and I cherry picked in the commits from #2909.
> 
> The v20 flavor of the binary I produced runs, though the node only has blocks from 6711001 onward, which is after the v19 upgrade and presumably is missing knowledge of the blocks which contains the messages we need to test.
> 
> If needed, we can run this v20 binary with a copy of an archive node's data.
> 
> Were there any messages present in v19 but not v20 that could be used to test?

The test would be the same as on the initial issue https://github.com/zeta-chain/node/issues/2835

This PR should solve the error encountered during the query but it is on mainnnet.

If needed I can look into having an example to test with Athens3

**Comment by CryptoFewka** [comment]
The same query provided in the issue also fails when pointed at an archive node. It returns this response:
```
{"code":2,"message":"codespace sdk code 2: tx parse error: unable to resolve type URL /zetachain.zetacore.crosschain.MsgGasPriceVoter","details":[]}
```

**Comment by lumtis** [comment]
> This does seem to fix `MsgVoteGasPrice` (looked at block `6325419` right before the v19 upgrade on Athens) but `MsgGasPriceVoter` from block `4667204` does not work. I would expect the same to be true for Mainnet blocks created on a pre-v16 network.

Ok, wanted to check with the initial query of the issue first. `MsgGasPriceVoter` was also missing, added it here, if we can check again the query with the new version of the binary would be great!

I checked the previous breaking changes, we should not have other message used frequently that should have this compatibility issue

File: proto/zetachain/zetacore/crosschain/legacy_msgs.proto [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  string creator = 1;
+  int64 chain_id = 2;
+  uint64 nonce = 3;
+  string tx_hash = 4;
+  pkg.proofs.Proof proof = 5;
+  string block_hash = 6;
+  int64 tx_index = 7;
+}
+
+// legacy MsgAddInboundTracker
```
[the remark this answers is not in the document]
Comment (fbac) [reply proto/zetachain/zetacore/crosschain/legacy_msgs.proto:25]: Agree on adding a Legacy suffix and keep it everywhere for consistency.

**Review by fbac — DISMISSED** [review]

File: proto/zetachain/zetacore/crosschain/legacy_msgs.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply proto/zetachain/zetacore/crosschain/legacy_msgs.proto:25]: The codec register the message with `RegisterImplementations` using the message name. 
If we change the name, we will likely have the error in https://github.com/zeta-chain/node/issues/2835#issuecomment-2338482524

File: proto/zetachain/zetacore/crosschain/tx.proto [hunk 3]
```diff
@@ -119,7 +119,8 @@ message MsgVoteGasPrice {
 
   uint64 block_number = 4;
 
-  reserved 5; // deprecated `string supply`
+  // unused
```
[the remark this answers is not in the document]
Comment (lumtis) [reply proto/zetachain/zetacore/crosschain/tx.proto]: d919a0f7

File: x/crosschain/types/codec.go [hunk 4]
```diff
@@ -19,6 +19,14 @@ func RegisterCodec(cdc *codec.LegacyAmino) {
 	cdc.RegisterConcrete(&MsgUpdateTssAddress{}, "crosschain/UpdateTssAddress", nil)
 	cdc.RegisterConcrete(&MsgAbortStuckCCTX{}, "crosschain/AbortStuckCCTX", nil)
 	cdc.RegisterConcrete(&MsgUpdateRateLimiterFlags{}, "crosschain/UpdateRateLimiterFlags", nil)
+
+	// unused messages defined for backward compatibility
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/types/codec.go]: d919a0f7

**Comment by lumtis** [comment]
@skosito @fbac please check back

**Review by kingpinXD — APPROVED** [review]

**Pull request merged**